src/resolve/resolved-conf.c

```diff
@@ -227,6 +228,105 @@ int config_parse_search_domains(
         return 0;
 }
 
+int config_parse_dns_netservice_name(const char *unit, const char *filename, unsigned line, const char *section, unsigned section_line, const char *lvalue, int ltype, const char *rvalue, void *data, void *userdata) {
+        DnsNetservice *ns = userdata;
+        _cleanup_free_ char *instance_name = NULL;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(ns);
+
+        if (isempty(rvalue)) {
+                r = -EINVAL;
+                log_syntax(unit, LOG_ERR, filename, line, r, "Service instance name can't be empty. Ignoring.");
+                return r;
+        }
+
+        if (strstr(rvalue, "%h")) {
+                _cleanup_free_ char *hostname = NULL;
+
+                r = gethostname_strict(&hostname);
+                if (r < 0) {
+                        log_warning_errno(r, "Can't get hostname for network service: %m. Ignoring.");
+                        return r;
+                }
+                instance_name = strreplace(rvalue, "%h", hostname);
```

Comment:
please use the specifier logic for this (see src/shared/specifier.h) In fact it already can handle `%h` easily for you anyway. (and it can also resolve other stuff that might make sense in this context)